flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/source/enumerator/ContinuousSplitPlannerImpl.java

```diff
@@ -0,0 +1,184 @@
+    Preconditions.checkArgument(scanContext.startSnapshotId() == null,
+        "Can't set startSnapshotId in ScanContext for continuous enumerator");
+    Preconditions.checkArgument(scanContext.endSnapshotId() == null,
+        "Can't set endSnapshotId in ScanContext for continuous enumerator");
+  }
+
+  @Override
+  public ContinuousEnumerationResult planSplits(IcebergEnumeratorPosition lastPosition) {
+    table.refresh();
+    if (lastPosition != null) {
+      return discoverDeltaSplits(lastPosition);
+    } else {
+      return discoverInitialSplits();
+    }
+  }
+
+  /**
+   * Discover delta changes between @{code lastPosition} and current table snapshot
+   */
+  private ContinuousEnumerationResult discoverDeltaSplits(IcebergEnumeratorPosition lastPosition) {
+    // incremental discovery mode
+    Snapshot currentSnapshot = table.currentSnapshot();
+    if (currentSnapshot.snapshotId() == lastPosition.endSnapshotId()) {
+      LOG.info("Current table snapshot is already enumerated: {}", currentSnapshot.snapshotId());
+      return new ContinuousEnumerationResult(Collections.emptyList(), lastPosition);
+    } else {
+      ScanContext incrementalScan = scanContext
+          .copyWithAppendsBetween(lastPosition.endSnapshotId(), currentSnapshot.snapshotId());
+      LOG.info("Incremental scan: startSnapshotId = {}, endSnapshotId = {}",
+          incrementalScan.startSnapshotId(), incrementalScan.endSnapshotId());
+      List<IcebergSourceSplit> splits = FlinkSplitPlanner.planIcebergSourceSplits(table, incrementalScan, workerPool);
+      IcebergEnumeratorPosition position = IcebergEnumeratorPosition.builder()
+          .startSnapshotId(lastPosition.endSnapshotId())
+          .startSnapshotTimestampMs(lastPosition.endSnapshotTimestampMs())
+          .endSnapshotId(currentSnapshot.snapshotId())
+          .endSnapshotTimestampMs(currentSnapshot.timestampMillis())
+          .build();
+      LOG.info("Discovered {} splits from incremental scan: {}", splits.size(), position);
+      return new ContinuousEnumerationResult(splits, position);
+    }
+  }
+
+  /**
+   * Discovery initial set of splits based on {@link IcebergEnumeratorConfig.StartingStrategy}
+   */
+  private ContinuousEnumerationResult discoverInitialSplits() {
+    HistoryEntry startSnapshotEntry = getStartSnapshot(table, config);
+    LOG.info("get startSnapshotId {} based on starting strategy {}",
+        startSnapshotEntry.snapshotId(), config.startingStrategy());
+    List<IcebergSourceSplit> splits;
+    if (config.startingStrategy() ==
+        IcebergEnumeratorConfig.StartingStrategy.TABLE_SCAN_THEN_INCREMENTAL) {
+      // do a full table scan first
+      splits = FlinkSplitPlanner.planIcebergSourceSplits(table, scanContext, workerPool);
+      LOG.info("Discovered {} splits from initial full table scan with snapshotId {}",
+          splits.size(), startSnapshotEntry);
+    } else {
+      splits = Collections.emptyList();
```

Comment:
We should also return splits by scanContext#copyWithSnapshotId(startSnapshotEntry.snapshotId()) for strategy `EARLIEST_SNAPSHOT`

Reply:
Thanks, got it.

Reply:
i still have a litter concern, let's take an example:

1. Table T initially has some snapshots S1, S2 with data files F1-3;
2. Then table T append F4, F5 with S3, and F6 with S4;
3. Now S1, S2 are removed based on retention policy, and table T has snapshots S3, S4 with data files F1-6.

Now S3 is the earliest snapshot, right?
IIUC, the current implementation will scan only F6 and the following append files.

I think we should scan F4-6 for `EARLIEST_SNAPSHOT` strategy to correspond to `LATEST_SNAPSHOT` strategy.

WDYT, @stevenzwu

Reply:
yes, that is the current behavior for exclusive starting snapshot. Because EARLIEST_SNAPSHOT is S3, the incremental scan of (S3, S4], we will only scan F6 only (committed by S4).

if it is LATEST_SNAPSHOT (S4), there won't be any files from F1-F6. Only the files are committed by the next snapshot (S5) will be discovered.

@yittg you like to make the starting snapshot inclusive (vs exclusive that is currently implemented), right? It is doable, although a little more complexity. If that is what other people favor, we can make the change.

Reply:
@stevenzwu i don't mean to make the starting inclusive. It should be consistent with the other places in iceberg.

We can compare the similar strategy in kafka.
I mean , it should include the earliest by `earliest`, and it should include the next coming by `latest`.

Reply:
Discussed with @yittg offline
* agree that both earliest and latest should have the same inclusive/exclusive behavior for easier . In the above example from Yi, here are the expectations if we use the exclusive behavior. earliest (starting S3) should discovery only F6 from S4 via the increment scan of (S3, S4]. latest (starting S4) should discover new files committed by future S5 via incremental scan of (s4, s5].
* because the current FlinkSource (`StreamingMonitorFunction`) implements exclusive behavior for `startSnapshotId` (probably matching the Iceberg `appendBetween` behavior), I am favoring keeping the exclusive behavior.

Will get some inputs from other people on the inclusive vs exclusive behavior regarding start snapshot.

Reply:
@yittg after discussing with @rdblue, we agreed on the inclusive starting behavior. With inclusive behavior, empty or non-empty tables should behave the same. For an empty table, earliest/latest will include all snapshots afterwards.

Reply:
Thanks for letting me know, good job.